File: initiator/steering.py
from collections import defaultdict
import inspect

MAX_EXTENT_SIZE = 128 * 1024
MIN_EXTENT_SIZE = 16 * 1024

steering_debug_on = False
def steering_debug(*args):
    if steering_debug_on:
        func_name = inspect.currentframe().f_back.f_code.co_name
        print("[DEBUG] Steering: {}: {}".format(func_name, " ".join(map(str, args))))

def EXTENT_ID(index):
    return (index << 12) // (MAX_EXTENT_SIZE)

class Steering:
    def __init__(self):
        print("[INFO] steering is on...")
        self.dict = defaultdict(None)
    
    """ Sorted by inode and by index """
    def sort(self, input_list):
        return sorted(input_list, key=lambda x: (x.ino, EXTENT_ID(x.index)))
    
    """ 
    If each inode is within a certain length,
    Returns the smallest page info.
    """
    def classify(self, input_list):
        cur_inode = input_list[0].ino
        cur_extent_id = EXTENT_ID(input_list[0].index)
        count = 1

        for i in range(1, len(input_list)):
            extent_id = EXTENT_ID(input_list[i].index)
            
            if input_list[i].ino == cur_inode and extent_id == cur_extent_id:
                count += 1
            else:
                adjacent = False
                if count > 1:
                    adjacent = True
                    #print(f"inode={cur_inode}, {cur_extent_id}S")
                else:
                    #print(f"inode={cur_inode}, {cur_extent_id}M")
                    pass

                self.dict[(cur_inode, cur_extent_id)] = adjacent

                cur_inode = input_list[i].ino
                cur_extent_id = extent_id
                count = 1

        # Print the last element
        adjacent = False
        if count > 1:
            adjacent = True
            #print(f"inode={cur_inode}, {cur_extent_id}S")
        else:
            #print(f"inode={cur_inode}, {cur_extent_id}M")
            pass

        self.dict[(cur_inode, cur_extent_id)] = adjacent

    def filter(self, input_list):
        used = (None, None)
        ret_list = []
        
        #print(self.dict)
        for x in input_list:
            inode = x.ino
            extent_id = EXTENT_ID(x.index)

            if self.dict[(inode, extent_id)] == True:
                if used == (inode, extent_id):
                    continue
                else:
                    if used[0] is not None:
                        del self.dict[used]
                    used = (inode, extent_id)
                    ret_list.append(x)
                    #print(f"{x} => Storage")
            else:
                del self.dict[(inode, extent_id)]
                #print(f"{x} => Memory")
        
        if used[0] is not None:
            del self.dict[used]
        
        return ret_list
    
    def filter_by_inode(self, input_list):
        inode_dict = {}
        ret_list = []

        for x in input_list:
            inode = x.ino
            extent_id = EXTENT_ID(x.index)

            if self.dict.get((inode, extent_id), False):
                if inode not in inode_dict:
                    inode_dict[inode] = (inode, extent_id)
                    ret_list.append([x])
                else:
                    inode_dict[inode] = (inode, extent_id)
                    ret_list[-1].append(x)
            else:
                del self.dict[(inode, extent_id)]
        return ret_list

    def process(self, msg_list):
        sorted_input_list = self.sort(msg_list)
        self.classify(sorted_input_list)
        ret_list = self.filter(sorted_input_list)
        
        steering_debug(ret_list)
        
        return ret_list
        

from my_types import *
if __name__ == "__main__":
    steering = Steering()
    
    dev_id = 0
    file_size = 0
    
    input_list = [
            PageInfo(EventType.EVENT_PAGE_DELETION, dev_id, 1, 100, file_size),
            PageInfo(EventType.EVENT_PAGE_DELETION, dev_id, 1, 50, file_size),
            PageInfo(EventType.EVENT_PAGE_DELETION, dev_id, 1, 10, file_size),
            ]
    ret_list = steering.process(input_list)
    print("=== Storage: tc: same inode, discontiguous index  ===")
    print(ret_list)

    input_list = [
            PageInfo(EventType.EVENT_PAGE_DELETION, dev_id, 5, 1, file_size),
            PageInfo(EventType.EVENT_PAGE_DELETION, dev_id, 1, 2, file_size),
            PageInfo(EventType.EVENT_PAGE_DELETION, dev_id, 4, 3, file_size),
            ]
    ret_list = steering.process(input_list)
    print("=== Storage: tc: different inode, adjacent index  ===")
    print(ret_list)

    input_list = [
            PageInfo(EventType.EVENT_PAGE_DELETION, dev_id, 5, 3, file_size),
            PageInfo(EventType.EVENT_PAGE_DELETION, dev_id, 1, 1, file_size),
            PageInfo(EventType.EVENT_PAGE_DELETION, dev_id, 4, 0, file_size),
            PageInfo(EventType.EVENT_PAGE_DELETION, dev_id, 1, 2, file_size),
            PageInfo(EventType.EVENT_PAGE_DELETION, dev_id, 5, 5, file_size),
            PageInfo(EventType.EVENT_PAGE_DELETION, dev_id, 5, 4, file_size),
            ]
    ret_list = steering.process(input_list)
    print("=== Storage ===")
    print(ret_list)

    input_list = [
            PageInfo(EventType.EVENT_PAGE_DELETION, dev_id, 1, 3, file_size),
            PageInfo(EventType.EVENT_PAGE_DELETION, dev_id, 1, 5, file_size),
            PageInfo(EventType.EVENT_PAGE_DELETION, dev_id, 1, 7, file_size),
            PageInfo(EventType.EVENT_PAGE_DELETION, dev_id, 1, 6, file_size),
            PageInfo(EventType.EVENT_PAGE_DELETION, dev_id, 1, 10, file_size),
            PageInfo(EventType.EVENT_PAGE_DELETION, dev_id, 1, 12, file_size),
            ]
    ret_list = steering.process(input_list)
    print("=== Storage ===")
    print(ret_list)

File: initiator-cpp/steering.cpp
#include <iostream>
#include <algorithm>

#include "steering.h"

#define MAX_EXTENT_SIZE (128 * 1024)

#define DMFP_STEERING_INFO
#ifdef DMFP_TRANS_INFO
#define dmfp_steering_info(str, ...) \
    printf("[INFO] %s: " str, __func__, ##__VA_ARGS__)
#else
#define dmfp_steering_info(str, ...) do {} while (0)
#endif

//#define DMFP_STEERING_DEBUG
#ifdef DMFP_TRANS_DEBUG
#define dmfp_steering_debug(str, ...) \
    printf("[DEBUG] %s: " str, __func__, ##__VA_ARGS__)
#else
#define dmfp_steering_debug(str, ...) do {} while (0)
#endif

Steering steering;

static inline int EXTENT_ID(const uint64_t index) {
	return (index << PAGE_SHIFT) / MAX_EXTENT_SIZE;
}

struct ComparePageDeletionEvent {
	bool operator()(const PageDeletionEvent& a, const PageDeletionEvent& b) const {
        int a_extent_id = EXTENT_ID(a.index);
        int b_extent_id = EXTENT_ID(b.index);
		return std::tie(a.ino, a_extent_id) < std::tie(b.ino, b_extent_id);
	}
};

Steering::Steering() {
    dmfp_steering_info("");
}

std::vector<PageDeletionEvent> Steering::Sort(const std::vector<PageDeletionEvent>& input_list) {
    std::vector<PageDeletionEvent> sorted_list = input_list;
    std::sort(sorted_list.begin(), sorted_list.end(), ComparePageDeletionEvent());
    return sorted_list;
}

void Steering::Classify(const std::vector<PageDeletionEvent>& input_list) {
	uint64_t cur_inode = input_list[0].ino;
	uint64_t cur_extent_id = EXTENT_ID(input_list[0].index);
	int count = 1;

	for (size_t i = 1; i < input_list.size(); ++i) {
		uint64_t extent_id = EXTENT_ID(input_list[i].index);

		if (input_list[i].ino == cur_inode && extent_id == cur_extent_id) {
			count += 1;
		} else {
			bool adjacent = false;
			if (count > 1) {
				adjacent = true;
			}

			dict[std::make_pair(cur_inode, cur_extent_id)] = adjacent;

			cur_inode = input_list[i].ino;
			cur_extent_id = extent_id;
			count = 1;
		}
	}

	// Print the last element
	bool adjacent = false;
	if (count > 1) {
		adjacent = true;
	}

	dict[std::make_pair(cur_inode, cur_extent_id)] = adjacent;
}

std::vector<PageDeletionEvent> 
Steering::Filter(const std::vector<PageDeletionEvent>& input_list) {
	std::pair<uint64_t, uint64_t> used = std::make_pair(-1, -1);
	std::vector<PageDeletionEvent> ret_list;

	for (const auto& x : input_list) {
		uint64_t inode = x.ino;
		uint64_t extent_id = EXTENT_ID(x.index);

		if (dict[std::make_pair(inode, extent_id)]) {
			if (used == std::make_pair(inode, extent_id)) {
				continue;
			} else {
				if (used.first != -1) {
					dict.erase(used);
				}
				used = std::make_pair(inode, extent_id);
				ret_list.push_back(x);
			}
		} else {
			dict.erase(std::make_pair(inode, extent_id));
		}
	}

	if (used.first != -1) {
		dict.erase(used);
	}

	return ret_list;
}

#if 0
std::vector<std::vector<PageDeletionEvent>> 
Steering::FilterByInode(const std::vector<PageDeletionEvent>& input_list) {
	std::map<uint64_t, std::pair<uint64_t, uint64_t>> inode_dict;
	std::vector<std::vector<PageDeletionEvent>> ret_list;

	for (const auto& x : input_list) {
		uint64_t inode = x.ino;
		uint64_t extent_id = EXTENT_ID(x.index);

		if (dict[std::make_pair(inode, extent_id)]) {
			auto it = inode_dict.find(inode);
			if (it == inode_dict.end()) {
				inode_dict[inode] = std::make_pair(inode, extent_id);
				ret_list.push_back({x});
			} else {
				inode_dict[inode] = std::make_pair(inode, extent_id);
				ret_list.back().push_back(x);
			}
		} else {
			dict.erase(std::make_pair(inode, extent_id));
		}
	}

	return ret_list;
}
#endif 

std::vector<PageDeletionEvent> 
Steering::Process(const std::vector<PageDeletionEvent> &msg_list) {
    std::vector<PageDeletionEvent> sorted_list = Sort(msg_list);
	Classify(sorted_list);
	std::vector<PageDeletionEvent> ret_list = Filter(sorted_list);
	return ret_list;
}


#if 0
int main(void) {
    std::vector<PageDeletionEvent> vec;

    uint32_t dev_id = 100;
    uint64_t file_size = 1024;

    PageDeletionEvent event = { dev_id, 5, 3, file_size }; 
    vec.push_back(event);
    PageDeletionEvent event2 = { dev_id, 1, 1, file_size }; 
    vec.push_back(event2);
    PageDeletionEvent event3 = { dev_id, 4, 0, file_size }; 
    vec.push_back(event3);
    PageDeletionEvent event4 = { dev_id, 1, 2, file_size }; 
    vec.push_back(event4);
    PageDeletionEvent event5 = { dev_id, 5, 5, file_size }; 
    vec.push_back(event5);
    PageDeletionEvent event6 = { dev_id, 5, 4, file_size }; 
    vec.push_back(event6);
    
    printf("vec cap=%lu, size=%lu\n", vec.capacity(), vec.size());
    vec = steering.Process(vec);
    
    for (auto &e : vec) {
        printf("(%lu, %lu) ", e.ino, e.index);
    }
    printf("\n");

    vec.clear();
    printf("vec cap=%lu, size=%lu\n", vec.capacity(), vec.size());
}
#endif
